### ISSUE TITLE
Add support for windows auth in xplat

### ISSUE THREAD

**Issue by emgarten** [NuGet/Home#2393]
NuGet xplat does not support windows auth. In some scenarios a null reference exception is hit when restore tries to authenticate against a source with windows auth.

https://github.com/dotnet/cli/issues/1679

**Comment by tebeco** [comment]
So this issue is still opened ?

**Comment by raffaeler** [comment]
Any workaround?
Looks like nuget never liked proxies, but this issue is blocking

**Comment by emgarten** [comment]
The best workaround is to use NuGet.exe 3.5.0-beta from http://dist.nuget.org/index.html

_dotnet restore_ is a subset of the restore command in NuGet.exe

**Comment by tebeco** [comment]
I could make 3.5.0beta works behind a proxy with ntlm auth

dotnet restore still fail

Vs2015 use cache inside program files

I have to manually run nuget.exe

So VsCode won't work because it use a separate nuget client

**Comment by raffaeler** [comment]
I need making dotnet work to avoid a continuous mess in VS2015.
From this perspective, the best workaround I found is installing cntlm proxy which works also with nuget 3.4 (the default for the current tools).

I am quite surprised the proxy problem is back since I raised this problem more than 2 years ago with the old nuget.exe.

**Comment by tebeco** [comment]
The pb is for nuget xplat is it the exact same codebase ?

**Comment by theBoringCoder** [comment]
Why is it that "dotnet new" works (in regards to retrieving packages) but Visual Studio 2015 Update 3 File --> New Project --> New .NET Core project fails to retrieve packages. Aren't they both using nugget.xplat?

**Comment by emgarten** [comment]
@tebeco the codebase for nuget.exe restore and dotnet restore are the same, nuget.exe is built against the desktop framework and contains msbuild and windows auth support. dotnet restore is built using the netstandard libraries and does not have support for all of these yet.

**Comment by alpaix** [comment]
Original dotnet/cli#1679 issue was caused by a bug in CoreFx (null pointer exception in `System.Net.Http`) of earlier versions. NuGet supports windows auth out of the box.

As we recently refactored our authentication stack I went ahead and validated reported scenario via _NuGet.Server v2.11.2.0_ configured with NTLM auth module on IIS Express. No issues spotted in:
- NuGet 3.5.0-rtm-1724
- .NET CLI 1.0.0-preview2-003121

```
.NET Command Line Tools (1.0.0-preview2-003121)

Product Information:
 Version:            1.0.0-preview2-003121
 Commit SHA-1 hash:  1e9d529bc5

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.14393
 OS Platform: Windows
 RID:         win10-x86
```
- .NET CLI 1.0.0-rc2-002543

```
.NET Command Line Tools (1.0.0-rc2-002543)

Product Information:
 Version:     1.0.0-rc2-002543
 Commit Sha:  38d0c28a1e

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.14393
 OS Platform: Windows
 RID:         win10-x64
```

**Comment by alpaix** [comment]
@raffaeler @tebeco Could you please confirm the issue has been resolved in your scenarios?

**Comment by raffaeler** [comment]
Unfortunately / luckily I can't because they finally put nuget in the proxy exception list

**Comment by tebeco** [comment]
@alpaix what have been resolved ?
and using what version of what ;)

I still have issue but it's really weird

On corporate network on .NetCore app

> dotnet.exe --version
> 1.0.0-preview2-003121

Nuget Path is in PATH var and it's CLI 3.5 beta2

> nuget
> NuGet Version: 3.5.0.1520

nuget VSIX is 3.5 Beta2

I have to run manually

> nuget restore

Every time I create a new project using for example NUnitit fails to restore pacakges
Unless I use nuget CLI :

> nuget restore

And it will restore packages from internet and not localCaches, i think the bug may comes from elsewhere

So i guess the NTLM issue is fixed in 3.5-beta2 if that was your question

Else i'm not sure to understand the question ;)

**Comment by alpaix** [comment]
@tebeco Few questions to you to clarify your complains:
- Do you know what server you use for package feed, e.g. NuGet.Server, Klondike, Visual Studio TFS? 
- Is your desktop behind corporate proxy?
- Does `dotnet restore` work?
- What exact VS version do you use, 2015 what update?
- What type of project do you create in VS when it fails restoring?
- Do you see any output messages in output window?

**Comment by tebeco** [comment]
- NuGet org api v3
- Ntlm proxy with ssl rewriting
- not in vs/vscode/console unless package are in ProgramFiles cache
- latest stable version of any tools / update / kb (vs2015.3 vscode dotnet is the preview from .net core 1.0 rtm)
- .net core project only (any kind) lib/console/mvc. The boilerplate work thx to ProgramFiles cache but package like xUnit / NSubstitute will fail to restore unless I use nuget cli 3.5beta2
- Error http 407 proxy with required

It's like nugey vsix (v3.5beta2 also) is either bugged or not used when I add manually a line in project.json and just save (not using/tried nuget gui or the powershell prompt inside studio)

And c# extension for vscode have the same issue when I try to open a folder of a netcore app, the extension is triggered but I got the same 407 in the console.
And for vscode I have no clues at all to know how to update the nugget client. It seems to be package within dotnet.exe that is within the c# extension (yep ... Inception)

All our corporate stcack in full of pain webapi so in q later future moving to .netcore is not that stupid and probably easy but not unless its tuned to work nicely with ntlm prosy ^^

**Comment by smezger** [comment]
@alpaix Thanks for solving this issue!
It's working now on my side behind the proxy with auth!
As you said, I've used:
-     NuGet 3.5.0-rtm-1724
-     .NET CLI 1.0.0-preview2-003121

**Comment by tebeco** [comment]
Out of curiosity, where does NuGet 3.5.0-rtm-1724 comes from ?
I only find beta2

**Comment by smezger** [comment]
You can download it here:
https://dist.nuget.org/win-x86-commandline/v3.5.0-rc1/NuGet.exe
However I’ve figured out that it is actually a newer build 1737 and not 1724.

**Comment by alpaix** [comment]
@tebeco @smezger We've just released 3.5 RC (see [the release notes](http://blog.nuget.org/20160811/Announcing-NuGet-3.5-RC.html)).

**Comment by alpaix** [comment]
@tebeco Can you share the output from:

```
dotnet restore ... --verbosity Debug
```

**Comment by Madlord** [comment]
Still got error 407 with actual 3.5rc installed.

**Comment by tebeco** [comment]
@alpaix
I'll give it a try today
I could not test earlier

**Comment by mtttcgcg** [comment]
Where should I place this new Nuget.exe in order to make VsCode's dotnet restore command use it?

**Comment by tebeco** [comment]
You cant
It's part of the c# extension if i'm not wrong

**Comment by alpaix** [comment]
@Madlord Would you be able to capture verbose output and/or Fiddler trace and share it? Please send it to support@nuget.org and mention issue id in the subject.

**Comment by alpaix** [comment]
@mtttcgcg VSCode invokes `dotnet restore` to perform NuGet package restore operations. In its turn it doesn't call NuGet.exe but rather uses _embedded_ NuGet API. You can install the latest .NET Core on your desktop from [here](https://microsoft.com/net/core).

**Comment by Barryrowe** [comment]
Has anyone confirmed a version of the nuget extension working through an NTLM proxy with a .net core project?

A fresh install of the .net core tools for VS2015  (Preview 2.0.1) shows the same issues. The nuget extenstion installed on my machine is:  3.5.0.1484, and was installed today.

The new dotnet core toolchain fails to properly authenticate to an NTML proxy  

In the same solution if we add an old-style project  (ASP.NET 4 for example), nuget packages install fine because, as is stated above, it's using the bundled nuget.exe and not the xplat version (or whatever the right name for that is)  as @brthor points out here https://github.com/dotnet/cli/issues/3065#issuecomment-224002424

_example error messages_

> C:\Program Files\dotnet\dotnet.exe restore "D:\git-repos\TestNugetInstall\ASPTEST.vs\restore.dg"
> log  : Restoring packages for D:\git-repos\TestNugetInstall\CoreNugetTest\project.json...
> log  : Retrying 'FindPackagesByIdAsyncCore' for source 'https://www.nuget.org/api/v2/FindPackagesById()?id='Microsoft.Web.Xdt''.
> log  : Response status code does not indicate success: 407 (Proxy Authentication Required).
> log  : Retrying 'FindPackagesByIdAsyncCore' for source 'https://www.nuget.org/api/v2/FindPackagesById()?id='Microsoft.Web.Xdt''.
> log  : Response status code does not indicate success: 407 (Proxy Authentication Required).
> error: Failed to retrieve information from remote source 'https://www.nuget.org/api/v2/FindPackagesById()?id='Microsoft.Web.Xdt''.
> error: Response status code does not indicate success: 407 (Proxy Authentication Required).
> error: Failed to retrieve information from remote source 'https://www.nuget.org/api/v2/FindPackagesById()?id='Microsoft.Web.Xdt''.
> error:   Response status code does not indicate success: 407 (Proxy Authentication Required).

**Comment by tebeco** [comment]
If you can provide me an url of the nuget.exe cli i'll try monday at work
A full url of the exe ^^

**Comment by tebeco** [comment]
So far nuget restore works but not dotnet restore
So maybe the version of dotnet could help too

**Comment by Barryrowe** [comment]
We are just using the bundled version with visual studio 2015 update 3, with the above .net core tools package and nuget extension installed.

The nuget extenstion installed on my machine is: 3.5.0.1484

We are not directly calling the cli, that is happening through the install process with a .net core based project in Visual Studio 2015. The error lines above show that this underneath calls dotnet restore.

The version of the dotnet CLI tools installed are:  .NET Command Line Tools (1.0.0-preview2-003121)

Just playing around trying to get you more information, it looks like if I set the http_proxy environment variable, the restore succeeds from the command line. 

AFTER running this from the command line, visual studio was able to properly restore as well, as long as the http_proxy environment variable is set. Setting the http_proxy environment variable did not have any effect on running the restore through Visual Studio until after I ran a command line restore.  (This is the most surprising find to me).

It appears that maybe whatever interface to the nuget api dotnet restore is using (.xplat?? I'm not familiar enough with the available interfaces to know), is only looking at the environment variable for a proxy url, where I'm guessing using the Nuget.exe restore, it checks the system and ensures the proxy variable is setup in the executing context before running the actual restore. This is just a guess, but hopefully this information helps.

This is being run on Windows 7 Professional

``` bash
C:\>set http_proxy=[myCorporateProxyWithoutCredentials]

C:\>dotnet -v restore "D:\git-repos\TestNugetInstall\ASPTEST\.vs\restore.dg"
Telemetry is: Enabled
log  : Restoring packages for D:\git-repos\TestNugetInstall\CoreNugetTest\project.json...
log  : Installing Newtonsoft.Json 8.0.1.
log  : Restoring packages for tool 'Microsoft.AspNetCore.Server.IISIntegration.Tools' in D:\git-repos\TestNugetInstall\CoreNugetTest\project.json...
log  : Writing lock file to disk. Path: D:\git-repos\TestNugetInstall\CoreNugetTest\project.lock.json
log  : D:\git-repos\TestNugetInstall\CoreNugetTest\CoreNugetTest.xproj
log  : Restore completed in 3340ms.


C:\>set http_proxy=
C:\>echo %http_proxy%
%http_proxy%

C:\>dotnet -v restore "D:\git-repos\TestNugetInstall\ASPTEST\.vs\restore.dg"
Telemetry is: Enabled
log  : Restoring packages for D:\git-repos\TestNugetInstall\CoreNugetTest\project.json...
log  : Retrying 'FindPackagesByIdAsyncCore' for source 'https://www.nuget.org/api/v2/FindPackagesById()?id='Newtonsoft.Json''.
log  : Response status code does not indicate success: 407 (Proxy Authentication Required).
log  : Retrying 'FindPackagesByIdAsyncCore' for source 'https://www.nuget.org/api/v2/FindPackagesById()?id='Newtonsoft.Json''.
log  : Response status code does not indicate success: 407 (Proxy Authentication Required).
error: Failed to retrieve information from remote source 'https://www.nuget.org/api/v2/FindPackagesById()?id='Newtonsoft.Json''.
error: Response status code does not indicate success: 407 (Proxy Authentication Required).
error: Failed to retrieve information from remote source 'https://www.nuget.org/api/v2/FindPackagesById()?id='Newtonsoft.Json''.
error:   Response status code does not indicate success: 407 (Proxy Authentication Required).
```

**Comment by tebeco** [comment]
# Versions

> ### nuget
> 
> NuGet Version: 3.5.0.1737
> 
> ### nuget update -self
> 
> Checking for updates from https://www.nuget.org/api/v2/.
> Currently running NuGet.exe 3.5.0-rtm-1737.
> NuGet.exe is up to date.
> 
> ### dotnet --version
> 
> 1.0.0-preview2-003131

# Project

Working with .Net Core App
Add a nuget that you don't have in Visual Studio Cache (I've used NUnit 3.5) in project.json

# Restore

## Dotnet CLI

> ### dotnet restore
> 
> log  : Restoring packages for D:\Sources\Other\NetCoreConsoleapp\src\NetCoreConsoleapp\project.json...
> error: Unable to load the service index for source https://api.nuget.org/v3/index.json.
> error:   Response status code does not indicate success: 407 (Proxy Authentication Required).

## Nuget CLI

TL;DR ==> Working

> ### nuget restore
> 
> MSBuild auto-detection: using msbuild version '14.0' from 'C:\Program Files (x86)\MSBuild\14.0\bin'.
> Restoring packages for D:\Sources\Other\NetCoreConsoleapp\src\NetCoreConsoleapp\project.json...
> ...

**Comment by Barryrowe** [comment]
@tebeco Based on the comments on from @brthor here dotnet/cli#3065 (comment) the nuget CLI and the dotnet cli use a different entry point to the nuget api.

Based on my testing, this does seem like an issue with the Visual Studio Nuget Extension (or visual studio itself). I'm just trying to provide the information.

**Comment by tebeco** [comment]
My answner only use CLI
I've used visual studio on tu create a project template and closed it
Nuget.exe and dotnet.exe ran into a console prompt

**Comment by nkolev92** [comment]
Hey all, 
Apologies for the long turnaround on this, at this point I think most of the underlying framework issues have been fixed, I believe that it's safe to close this issue. 

Please file new issues for any new problems you might hit.
